packager/packager.cc

```diff
@@ -0,0 +1,665 @@
+#include "packager/packager.h"
+#include "absl/strings/match.h"
+#include "absl/strings/str_join.h"
+#include "absl/strings/str_replace.h"
+#include "ast/Helpers.h"
+#include "ast/treemap/treemap.h"
+#include "common/FileOps.h"
+#include "common/concurrency/ConcurrentQueue.h"
+#include "common/concurrency/WorkerPool.h"
+#include "common/sort.h"
+#include "core/Unfreeze.h"
+#include "core/errors/packager.h"
+
+using namespace std;
+
+namespace sorbet::packager {
+namespace {
+
+constexpr string_view PACKAGE_FILE_NAME = "__package.rb"sv;
+
+struct PackageName {
+    core::LocOffsets loc;
+    core::NameRef mangledName = core::NameRef::noName();
+    vector<core::NameRef> fullName;
+    ast::TreePtr getFullNameLiteral(core::LocOffsets loc) const;
+};
+
+struct PackageInfo {
+    // The path prefix before every file in the package, including path separator at end.
+    std::string packagePathPrefix;
+    PackageName name;
+    // loc for the package definition. Used for error messages.
+    core::Loc loc;
+    // The mangled names of each package imported by this package.
+    vector<PackageName> importedPackageNames;
+    // List of exported items that form the body of this package's public API.
+    // These are copied into every package that imports this package.
+    ast::ClassDef::RHS_store exportedItems;
+};
+
+class NameFormatter {
+    const core::GlobalState &gs;
+
+public:
+    NameFormatter(const core::GlobalState &gs) : gs(gs) {}
+
+    void operator()(std::string *out, core::NameRef name) const {
+        out->append(name.data(gs)->shortName(gs));
+    }
+};
+
+// Gets the package name in `name` if applicable.
+PackageName getPackageName(core::MutableContext ctx, ast::TreePtr &name, bool errorOnInvalidName = false) {
+    PackageName pName;
+    pName.loc = name->loc;
+
+    auto constLit = ast::cast_tree<ast::UnresolvedConstantLit>(name);
+    while (constLit != nullptr) {
+        if (errorOnInvalidName && absl::StrContains(constLit->cnst.data(ctx)->shortName(ctx), "_")) {
```

Comment:
I'm usually skeptical of optional `bool` arguments. Doubly so when the function is only called once with it `true` and once with it `false`. That almost always indicates to me that the function is two separate functions in disguise.

Suggestion: factor out the error handling code, and make it a precondition to calling this function.

Reply:
Split into 2 functions, and cleaned it up so they are passed an `UnresolvedConstantLit`. As a nice bonus, `mangledName` will now always exist.